torch/distributed/__init__.py

```diff
@@ -15,8 +16,11 @@ def is_available():
     return hasattr(torch._C, "_c10d_init")
 
 
-if is_available() and not torch._C._c10d_init():
-    raise RuntimeError("Failed to initialize torch.distributed")
+if platform in ["linux", "darwin"]:
+    if not (is_available() and torch._C._c10d_init()):
```

Comment:
Any reason for changing this condition? This means when PyTorch is compiled with `USE_DISTRIBUTED=0`, importing `torch.distributed` would throw an error, while it won't through before this change until the application try to call some API from `torch.distributed`?

Reply:
In `USE_DISTRIBUTED=0` compiled pytorch version, importing `torch.distributed` raised "ModuleNotFoundError" / issue: https://github.com/pytorch/pytorch/issues/40555 - `is_available()` is `false` so `RuntimeError` wouldn't call.
IMO, it can be confusing & misleading user behavior without warning message

Reply:
```python
>>> import torch.distributed as dist
>>> dist.is_available()
False
>>> import torch.distributed.ReduceOp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'torch.distributed.ReduceOp'
>>> dir(dist)
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'absolute_import', 'autograd', 'division', 'is_available', 'print_function', 'rpc', 'torch', 'unicode_literals']
```

Above is the current behavior, when importing `torch.distributed`, it won't break, but importing modules or functions in `torch.distributed` will, because they are not available.

The current solution in this PR will change this behavior and can potentially break applications. Say if an application does sth like below, the code now would then crash.

```python
import torch.distributed

if torch.distributed.is_available():
    # do something
else:
    # do something else
```

If this breakage is expected, we will need to mark this PR as [BC-breaking] and then draw in more people for discussion. Let me know which one you prefer.

Reply:
As you mentioned, I looked up `distributed.is_available` has used a lot in GitHub repos, so change current behavior will cause a dependency problem. It seems better to leave it as it is. Thank you for reviewing pr